ldif.go

```diff
@@ -0,0 +1,256 @@
+				line = []byte{}
+				lines = [][]byte{}
+				if err == io.EOF {
+					return nil
+				}
+			default:
+				switch nextLine[0] {
+				case Comment:
+					continue
+				case SPACE:
+					if len(nextLine) > 1 {
+						line = append(line, nextLine[1:]...)
+						continue
+					} else {
+						return l.newError("space only line")
+					}
+				default:
+					if len(line) != 0 {
+						lines = append(lines, line)
+					}
+					line = nextLine
+					continue
+				}
+			}
+		default:
+			return l.newError(err.Error())
+		}
+	}
+	if len(lines) != 0 {
+		entry, perr := l.parseEntry(lines)
+		if perr != nil {
+			return l.newError(perr.Error())
+		}
+		l.Entries = append(l.Entries, entry)
+	}
+	return nil
+}
+
+func (l *LDIF) parseEntry(lines [][]byte) (entry *Entry, err error) {
+	// for i, line := range lines {
+	//     fmt.Fprintf(os.Stderr, "% 2d %s\n", i, line)
+	// }
+	if l.version == 0 && bytes.HasPrefix(lines[0], []byte("version:")) {
+		line := bytes.TrimLeft(lines[0][8:], SPACES)
+		if version, err := strconv.Atoi(string(line)); err != nil {
+			return nil, err
+		} else {
+			if version != 1 {
+				return nil, errors.New("Invalid version spec " + string(line))
+			}
+			l.version = 1
+			lines = lines[1:]
+		}
+	}
+
+	if !bytes.HasPrefix(lines[0], []byte("dn:")) {
+		return nil, errors.New("Missing dn:")
+	}
+	_, val, err := l.parseLine(lines[0])
+	if err != nil {
+		return nil, err
+	}
+	dn := val
+
+	lines = lines[1:]
+	if bytes.HasPrefix(lines[0], []byte("changetype:")) {
+		_, val, err := l.parseLine(lines[0])
+		if err != nil {
+			return nil, err
+		}
+		l.changeType = val
+		lines = lines[1:]
+	}
+	if l.changeType != "" {
+		return nil, errors.New("change records not supported")
+	}
+
+	attrs := make(map[string][]string)
+	for i := 0; i < len(lines); i++ {
+		attr, val, err := l.parseLine(lines[i])
+		if err != nil {
+			if !l.RelaxedParser {
+				return nil, err
+			} else {
+				continue
+			}
+		}
+		if _, ok := attrs[attr]; ok {
+			attrs[attr] = append(attrs[attr], string(val))
+		} else {
+			attrs[attr] = []string{string(val)}
+		}
+	}
+	return NewEntry(dn, attrs), nil
+}
+
+func (l *LDIF) parseLine(line []byte) (attr, val string, err error) {
+	off := 0
+	for len(line) > off && line[off] != ':' {
+		off++
+		if off >= len(line) {
+			return
+		}
+	}
+	if off == len(line) {
+		err = errors.New("Missing : in line")
+		return
+	}
+	if off > len(line)-2 {
+		err = errors.New("empty value")
+		return
+	}
+
+	attr = string(line[0:off])
+	if err = validAttr(attr); err != nil {
+		attr = ""
+		val = ""
+		return
+	}
+
+	switch line[off+1] {
+	case ':':
+		var n int
+		value := bytes.TrimLeft(line[off+2:], SPACES)
+		// fmt.Fprintf(os.Stderr, "LINE=%s\nVALUE=%s\n", line, value)
+		dec := make([]byte, base64.StdEncoding.DecodedLen(len(value)))
+		n, err = base64.StdEncoding.Decode(dec, value)
+		if err != nil {
+			return
+		}
+		val = string(dec[:n])
+	case '<': // FIXME missing return for *net.URL type
```

Comment:
should this return a `*net.URL` or inline the content located at the specified URL?

This makes it sound like an entry with `myattr:< file:///myfile.txt` should result in an entry where the `myattr` attribute is set to the content of `/myfile.txt`:

> 6) When a URL is specified in an attrval-spec, the following conventions apply:
> a) Implementations SHOULD support the file:// URL format.  The contents of the referenced file are to be included verbatim in the interpreted output of the LDIF file.
> b) Implementations MAY support other URL formats.  The semantics associated with each supported URL will be documented in an associated Applicability Statement.